### PR TITLE
Week-Selector, Month-Selector and Signature Canvas components 

### PR DESCRIPTION
**Pull request by apenalver2**

### CODE REVIEW

File: src/app/systelab-components/signature-canvas/signature-canvas.component.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    }
+  }
+  public DownloadImage(){
+    var canvas: HTMLCanvasElement = this.sign.nativeElement;
+    var dataURL = canvas.toDataURL('image/png');
+    this.href = dataURL;
+    var date = new Date();
+    this.downPath = date + "signature.png";
+  }
+  public CleanCanvas(){
```
Comment (alfonsserra) [inline src/app/systelab-components/signature-canvas/signature-canvas.component.ts]: Should start in lowercase

File: src/app/systelab-components/month-selector/month-selector.component.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  }
+  public hideSelector() {
+    this.isOpened = false;
+  }
+  public getMonths() {
+    this.months = [];
+    var idmonth = this.currentDate.getMonth();
+    var year = this.currentDate.getFullYear();
+    for (var i = 0; i < this.monthNames.length; i++) {
+      var d = new Date(year, i + 1, 1, 0, 0, 0, 0);
```
Comment (alfonsserra) [inline src/app/systelab-components/month-selector/month-selector.component.ts]: var should be let or const

File: src/app/systelab-components/week-selector/week-selector.component.ts [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    this.daysNames.push(this.i18nService.instant("COMMON_SUNDAY"));
+    this.daysNames.push(this.i18nService.instant("COMMON_MONDAY"));
+    this.daysNames.push(this.i18nService.instant("COMMON_TUESDAY"));
+    this.daysNames.push(this.i18nService.instant("COMMON_WEDNESDAY"));
+    this.daysNames.push(this.i18nService.instant("COMMON_THURSDAY"));
+    this.daysNames.push(this.i18nService.instant("COMMON_FRIDAY"));
+    this.daysNames.push(this.i18nService.instant("COMMON_SATURDAY"));
+  }
+
+  ngAfterViewInit(): void {
```
Comment (alfonsserra) [inline src/app/systelab-components/week-selector/week-selector.component.ts]: public

File: src/app/systelab-components/week-selector/week-selector.component.ts [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+  }
+  public onClicked(e) {
+    this.isOpened = true;
+  }
+  public hideSelector() {
+    this.isOpened = false;
+  }
+  public getMonths() {
+    this.months = [];
+    var idmonth = this.currentDate.getMonth();
```
Comment (alfonsserra) [inline src/app/systelab-components/week-selector/week-selector.component.ts]: var sould be let or const

File: src/app/systelab-components/week-selector/week-selector.component.ts [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+      }
+    }
+
+    this.getWeeks();
+  }
+  public getWeeks() {
+    this.weeks = [];
+    var monthWeeksActive = this.months[this.currentDate.getMonth()].month;
+    var yearActive = this.months[this.currentDate.getMonth()].year;
+    var g = this.getWeeksStartAndEndInMonth(monthWeeksActive, yearActive, 'm', 1);
```
Comment (alfonsserra) [inline src/app/systelab-components/week-selector/week-selector.component.ts]: g? Bad name for a variable

File: src/app/systelab-components/week-selector/week-selector.component.ts [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+    var week = this.getWeek(this.currentDate.getFullYear(), this.currentDate.getMonth(), this.currentDate.getDate());
+    for (var i = 0; i < g.length - 1; i++) {
+      var isActive = false;
+      var d = new Date(yearActive, monthWeeksActive, g[i].start, 0, 0, 0, 0);
```
Comment (alfonsserra) [inline src/app/systelab-components/week-selector/week-selector.component.ts]: d?

File: src/app/systelab-components/week-selector/week-selector.component.ts [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+      if (this.checkDateIntoIntervals(d)) {
+        var dend = new Date(yearActive, monthWeeksActive, g[i].end, 0, 0, 0, 0);
```
Comment (alfonsserra) [inline src/app/systelab-components/week-selector/week-selector.component.ts]: dend?

File: src/app/systelab-components/week-selector/week-selector.component.ts [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+    this.currentDate = new Date(week.year, week.month, week.day);
+    this.hideSelector();
+    this.getWeeks();
+  }
+  public onClickedMonth(month: Month) {
+    this.currentDate = new Date(month.year, month.month, 1);
+    this.getMonths();
+  }
+  public changeYear(v: number) {
+    var da = this.currentDate;
```
Comment (alfonsserra) [inline src/app/systelab-components/week-selector/week-selector.component.ts]: rename variables

File: src/app/systelab-components/week-selector/week-selector.component.ts [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+      }
+      else {
+        return false;
+      }
+    }
+    else {
+      return true;
+    }
+  }
+  public getWeeksStartAndEndInMonth(month: number, year: number, start: string, st: number) {
```
Comment (alfonsserra) [inline src/app/systelab-components/week-selector/week-selector.component.ts]: Too complex. split

File: src/app/systelab-components/week-selector/week-selector.component.ts [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+      end = st === 1 && end === 8 ? 1 : end;
+      if (end > numDays) {
+        end = numDays;
+      }
+      w.push({ start: st, end: end, number: this.getWeek(year, month, st) });
+    }
+    return w;
+  }
+
+  public getWeek(year, month, day) {
```
Comment (alfonsserra) [inline src/app/systelab-components/week-selector/week-selector.component.ts]: move to systelab-translate?

File: src/app/systelab-components/week-selector/week-selector.component.ts [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+    month += 1;//use 1-12
+    var a = Math.floor((14 - (month)) / 12);
```
Comment (alfonsserra) [inline src/app/systelab-components/week-selector/week-selector.component.ts]: use let not var

File: src/app/systelab-components/week-selector/week-selector.component.ts [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+    var y = year + 4800 - a;
```
Comment (alfonsserra) [inline src/app/systelab-components/week-selector/week-selector.component.ts]: rename variables

File: src/app/showcase/components/datepicker/showcase-datepicker.component.html [hunk 13]
```diff
@@ -31,4 +31,19 @@
                     [(currentDate)]="myDate"></systelab-datepicker>
         </div>
     </div>
+    <div class="row mt-1">
+        <label class="col-md-3 col-form-label" for="form-h-s">Date required bottom</label>
+        <div class="col-md-4">
+            <systelab-month-selector #wselectorMonth [isDisabled]="isDisabled" [maxDate]="maxDate" [minDate]="minDate" [currentDate]="currentDate"
```
Comment (alfonsserra) [inline src/app/showcase/components/datepicker/showcase-datepicker.component.html]: #w? And do you really need the reference?

File: src/app/showcase/components/datepicker/showcase-datepicker.component.html [hunk 14]
```diff
@@ -31,4 +31,19 @@
...
+            [(selectedMonth)]="selectedMonth"></systelab-month-selector>
+        </div>
+    </div>
+    <div class="row mt-1">
+        <label class="col-md-3 col-form-label" for="form-h-s">Date required bottom</label>
+        <div class="col-md-7">
+            <systelab-week-selector #wselectorWeek [isDisabled]="isDisabled" [maxDate]="maxDate" [minDate]="minDate" [currentDate]="currentDate"
```
Comment (alfonsserra) [inline src/app/showcase/components/datepicker/showcase-datepicker.component.html]: Same comment

File: src/app/showcase/components/datepicker/showcase-datepicker.component.ts [hunk 15]
```diff
@@ -1,13 +1,23 @@
...
 	selector:    'showcase-datepicker',
 	templateUrl: 'showcase-datepicker.component.html'
 })
 export class ShowcaseDatepickerComponent {
 
 	public myDate = new Date();
-
+	public maxDate: Date;
+	public minDate: Date;
+	public currentDate: Date;
```
Comment (alfonsserra) [inline src/app/showcase/components/datepicker/showcase-datepicker.component.ts]: So confuse to have myDate and currentDate

File: src/app/showcase/components/signature-canvas/showcase-signature-canvas.component.ts [hunk 16]
```diff
@@ -0,0 +1,20 @@
+import {Component} from '@angular/core';
+
+@Component({
+  selector: 'showcase-signature-canvas',
+  templateUrl: './showcase-signature-canvas.component.html'
+})
+export class ShowcaseSignatureCanvasComponent {
+  public withCodeVerification: boolean = true;
+  public height: number = 200;
```
Comment (alfonsserra) [inline src/app/showcase/components/signature-canvas/showcase-signature-canvas.component.ts]: No need for the type

File: src/app/systelab-components/month-selector/month-selector.component.ts [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+        const monthObj: Month = new Month(i, this.monthNames[i], year, isActive);
+        this.months.push(monthObj);
+        if (isActive) {
+          this.selectedMonth = monthObj;
+        }
+      }
+    }
+    console.log(this.months);
+  }
+  public changeYear(v: number) {
```
Comment (alfonsserra) [inline src/app/systelab-components/month-selector/month-selector.component.ts]: changeYear of addYear?

File: src/app/systelab-components/month-selector/month-selector.component.ts [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+    this.isOpened = false;
+  }
+  public getMonths() {
+    this.months = [];
+    const idmonth = this.currentDate.getMonth();
+    const year = this.currentDate.getFullYear();
+    for (let i = 0; i < this.monthNames.length; i++) {
+      const d = new Date(year, i + 1, 1, 0, 0, 0, 0);
+      if (this.checkDateIntoIntervals(d)) {
+        const isActive = i === idmonth ? true : false;
```
Comment (alfonsserra) [inline src/app/systelab-components/month-selector/month-selector.component.ts]: ????

const isActive = (i === idmonth);

File: src/app/systelab-components/month-selector/month-selector.component.ts [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+  }
+  public onClicked(e) {
+    this.isOpened = true;
+  }
+  public hideSelector() {
+    this.isOpened = false;
+  }
+  public getMonths() {
+    this.months = [];
+    const idmonth = this.currentDate.getMonth();
```
Comment (alfonsserra) [inline src/app/systelab-components/month-selector/month-selector.component.ts]: idmonth?

File: src/app/systelab-components/month-selector/month-selector.component.ts [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+    const year = this.currentDate.getFullYear();
+    for (let i = 0; i < this.monthNames.length; i++) {
+      const d = new Date(year, i + 1, 1, 0, 0, 0, 0);
```
Comment (alfonsserra) [inline src/app/systelab-components/month-selector/month-selector.component.ts]: d?

File: src/app/systelab-components/month-selector/month-selector.component.ts [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+        if (isActive) {
+          this.selectedMonth = monthObj;
+        }
+      }
+    }
+    console.log(this.months);
+  }
+  public changeYear(v: number) {
+    const da = this.currentDate;
+    const curr_year = da.getFullYear() + (v);
```
Comment (alfonsserra) [inline src/app/systelab-components/month-selector/month-selector.component.ts]: _? currentYear?

File: src/app/systelab-components/month-selector/month-selector.component.ts [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+    const gDate = new Date(curr_year, da.getMonth(), da.getDate());
```
Comment (alfonsserra) [inline src/app/systelab-components/month-selector/month-selector.component.ts]: g?

File: src/app/systelab-components/month-selector/month-selector.component.ts [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+    if (this.checkDateIntoIntervals(gDate)) {
+      this.currentDate = gDate;
+      this.setMonthsNames(this.currentDate.getFullYear());
+      this.getMonths();
+    }
+  }
+  public checkDateIntoIntervals(date: Date) {
+    if (this.maxDate && this.minDate) {
```
Comment (alfonsserra) [inline src/app/systelab-components/month-selector/month-selector.component.ts]: Change the order

File: src/app/systelab-components/styles/sass/_signature-canvas.scss [hunk 24]
```diff
@@ -0,0 +1,23 @@
...
+    & .alert{
+        padding: 0.5rem 0.3rem;
+    }
+    & .optionsBar{
+        margin-top:10px;
+    }
+    & .slab-form{
+        padding:15px;
+    }
+    & .mt-15{
```
Comment (alfonsserra) [inline src/app/systelab-components/styles/sass/_signature-canvas.scss]: NEVER DO THIS!!!!!!!

File: src/app/systelab-components/week-selector/week-selector.component.ts [hunk 25]
```diff
@@ -0,0 +1,199 @@
...
+  public onClickedWeek(week: Week) {
+    this.currentDate = new Date(week.year, week.month, week.day);
+    this.hideSelector();
+    this.getWeeks();
+  }
+  public onClickedMonth(month: Month) {
+    this.currentDate = new Date(month.year, month.month, 1);
+    this.getMonths();
+  }
+  public changeYear(v: number) {
```
Comment (alfonsserra) [inline src/app/systelab-components/week-selector/week-selector.component.ts]: change or add??

File: src/app/systelab-components/week-selector/week-selector.component.ts [hunk 26]
```diff
@@ -0,0 +1,199 @@
...
+    const curr_year = this.currentDate.getFullYear() + (v);
```
Comment (alfonsserra) [inline src/app/systelab-components/week-selector/week-selector.component.ts]: _??????

File: src/app/systelab-components/week-selector/week-selector.component.ts [hunk 27]
```diff
@@ -0,0 +1,199 @@
...
+  public changeMonth(v: number) {
+    const curr_month = this.currentDate.getMonth() + (v);
+    const newDate = new Date(this.currentDate.getFullYear(), curr_month, this.currentDate.getDate());
+    if (this.checkDateIntoIntervals(newDate)) {
+      this.currentDate = newDate;
+      this.getMonths();
+    }
+  }
+  public checkDateIntoIntervals(date: Date) {
+    if (this.maxDate && this.minDate) {
```
Comment (alfonsserra) [inline src/app/systelab-components/week-selector/week-selector.component.ts]: Change the order

File: src/app/systelab-components/month-selector/month-selector.component.ts (hunk 17 above)
Comment (apenalver2) [reply src/app/systelab-components/month-selector/month-selector.component.ts]: This function is used to add and subtract years, as well as months (Is from the systelab-calendar-header), (previousYear)="changeYear(-1)"
        (nextYear)="changeYear(1)"

File: src/app/systelab-components/styles/sass/_signature-canvas.scss (hunk 24 above)
Comment (apenalver2) [reply src/app/systelab-components/styles/sass/_signature-canvas.scss]: What is not advisable? use this class key or to refer to the child items that have that class?

**Pull request merged**